### ISSUE TITLE
Linux system headers not detected as system headers

### ISSUE THREAD

**Issue by pudelkoM** [cpplint/cpplint#146]
PR #133 swapped the dynamic system header detection from `<>` brackets to a hard coded list of names. This causes false positives in files with Linux specific includes. Unless I'm mistaken and files under `/usr/include` should not be classified as system headers.

test.h
```c++
// Copyright 2020 Open Networking Foundation
// SPDX-License-Identifier: Apache-2.0

#ifndef _PRIVATE_TMP_TEST_H_
#define _PRIVATE_TMP_TEST_H_

#include <net/ethernet.h>

#include <vector>

#include "absl/base/thread_annotations.h"

#endif  // _PRIVATE_TMP_TEST_H_
```

```bash
# cpplint test.h
test.h:9:  Found C++ system header after other system header. Should be: test.h, c system, c++ system, other.  [build/include_order] [4]
Done processing test.h
Total errors found: 1
```

**Comment by tkruse** [comment]
Thanks for the report. @Igulich, any opinion? MAybe we can also add linux-specific system headers to mitigate.

**Comment by tkruse** [comment]
So in this case, the question is whether e.g. `#include <net/ethernet.h>` (and several others) is regarded a C system header for cpplint or not. I believe those files we need to talk about are all part of the glibc package (for linux or other specific OSes).
E.g. as in this RPM package: https://centos.pkgs.org/6/centos-x86_64/glibc-headers-2.12-1.212.el6.x86_64.rpm.html

This document does not include e.g `net/ethernet.h`: https://www.gnu.org/software/libc/manual/html_mono/libc.html#Library-Summary.

Generally I can understand the argument that all packaged glibc header files are considered to be as much "c system headers" as the documented C, POSIX and GNULib headers, the lack of documentation for those headers does not make them "not system", though they are a little "less standard", e.g. not available for Cygwin.

So I am not sure where to get the full list of glibc header files from, since their documentation does not seem to provide a list. But we can add a list of header files from package glibc, and then add more if users find we missed any.

**Comment by pudelkoM** [comment]
Thanks for the quick response!
Just tried your PR and it has nearly fixed the issue. Now I'm just having one failing across the code base:

```c
#include <linux/filter.h>
```

Add `|linux` here?
https://github.com/cpplint/cpplint/blob/efe2ac5be1f45976721073ec24a0d304dda2e7fa/cpplint.py#L4908

I guess there is some concern about maintaining an ever growing list of system dependent filenames.

**Comment by tkruse** [comment]
@pudelkoM , I don't have any practical issue with adding linux/ to the folders, but I would like to have some better justification for future cases,for documentation, and for checking right now what other folders should be added to the list (drm, sound, video, xen?).

As far as I can see, linux/filter.h is not part of glibc-dev, but instead part of kernel headers, e.g. https://rpmfind.net/linux/RPM/fedora/devel/rawhide/aarch64/k/kernel-headers-5.7.0-0.rc7.git0.1.fc33.aarch64.html

> kernel-headers includes the C header files that specify the interface between the Linux kernel and userspace libraries and programs. 

Which from the perspective of a linter tool, might not count as a C system header (or might). A base question is, should cpplint have any opinion on what C headers in /usr/include are "system headers" or not? According to #43, the set of C-headers in /usr/include can be split into "system C headers" and "non-system C headers".

The current google styleguide says little on the subject (https://google.github.io/styleguide/cppguide.html):

> C system headers (more precisely: headers in angle brackets with the .h extension), e.g., <unistd.h>, <stdlib.h>

So to the google styleguide, all includes in angle brackets with .h extension are assumed to be C system headers (not C++ headers). Maybe that's also a hard truth for google internally (maybe their environments are restricted that way).

GCC declares (https://gcc.gnu.org/onlinedocs/cpp/System-Headers.html):

> Normally, only the headers found in specific directories are considered system headers.

By which we might consider all or /usr/include as system headers, I guess. Without being able to tell C from C++ headers.

Let's look at an example that motivated #43:

```
#include "foo/server/fooserver.h"

#include <sys/types.h> // C standard library
#include <unistd.h>      // C standard library
#include <gtest/gtest.h>   // installed C++ library

#include <string>        // C++ system library
#include <vector>       // C++ system library

#include "base/basictypes.h"
```
This is the original cpplint-accepted way of adding `#include <gtest/gtest.h>` to other includes. And in #43 this is called out to be wrong, because gtest/gtest.h is not a C header, and the cpplint warning when moving `#include <gtest/gtest.h>` to the c++ includes is:

    "Found C system header after C++ system header...[build/include_order]"

So now I believe there are only a few ways to get a consistent behavior:

* Revert #133 and instead change the linter warning test to something like:
  "Found C-style system header include after C++-style system header include..."
* Decide on a generally acceptable standard for what is a C system header for cpplint, similar to the suggestion in #43 (if that's possible), which satisfies most users.
* Change the warning by cpplint to avoid the mention of "system header", and instead use "standard library header".

If it is not likely to find a general consensus by which some C headers in /usr/include are "system headers" and others are not, I think I'd need to revert #136 

The purpose of the linter ordering rule is to make files in source code readable. A typical reader should not be surprised by which line/section an include is added. The google cpplint behavior just groups all `<foo1.h>` includes, then `<foo2>`, then other `"foo3.h"`. It expresses the intention to separate C system headers from C++ system headers, even if the given behavior does not achieve that perfectly (at least not without additional constraint of never including a C++ header using the `<foo.h>` form).

I am a bit lost now and feel like reverting #133 to at least remain compliant with upstream cpplint.

**Comment by tkruse** [comment]
The fuchsia styleguide adds some more google perspective
https://fuchsia.googlesource.com/fuchsia/+/master/docs/development/languages/c-cpp/cpp-style.md :

> If the header being included is a system, global, or library header (see Naming C/C++ objects for precise definitions), use <angle brackets> and the complete name of the header. These headers are considered “C library headers” for the purposes of the Google C++ Style Guide:

So this seems to indicate that C++ headers are tolerated to be considered “C library headers” by he linter tool. It is not clear if that is considered ideal, or just the path of least resistance to have a linter.

**Comment by tkruse** [comment]
FWIW, this is also discussed here: https://github.com/isocpp/CppCoreGuidelines/pull/1456

**Comment by caioaamaral** [comment]
> FWIW, this is also discussed here: [isocpp/CppCoreGuidelines#1456](https://github.com/isocpp/CppCoreGuidelines/pull/1456)

this [one](https://github.com/isocpp/CppCoreGuidelines/pull/1456#issuecomment-565734871) sounds like a great argument for using `<>` (avoiding obscure errors if including wrong files).

There are more about this discussion here: isocpp/CppCoreGuidelines#1596

**Comment by pudelkoM** [comment]
Thanks for fixing this!
